src/checkpoint_db.rs

```diff
@@ -20,77 +20,90 @@ use exonum::storage::{Database, Patch, Result as StorageResult, Snapshot};
+        self.inner
+            .read()
+            .expect("Cannot lock CheckpointDb for snapshot")
+            .snapshot()
+    }
+
+    fn merge(&self, patch: Patch) -> StorageResult<()> {
+        self.inner
+            .write()
+            .expect("Cannot lock CheckpointDb for merge")
+            .merge(patch)
+    }
+
+    fn merge_sync(&self, patch: Patch) -> StorageResult<()> {
+        self.merge(patch)
     }
+}
+
+/// Handler to a checkpointed database.
+#[derive(Debug)]
+pub struct CheckpointDbHandler<T> {
+    inner: Arc<RwLock<CheckpointDbInner<T>>>,
+}
 
+impl<T: Database> CheckpointDbHandler<T> {
     /// Rolls back this database by udoing the latest `count` `merge()` operations.
     ///
     /// # Panics
     ///
     /// - Panics if more operations are attempted to be reverted than the number of operations
     ///   in the DB journal.
-    pub fn rollback(&mut self, count: usize) -> bool {
-        let journal_len = self.journal
-            .read()
-            .expect("Cannot acquire read lock on journal")
-            .len();
-        assert!(
-            journal_len >= count,
-            "Cannot rollback {} changes; only {} checkpoints in the journal",
-            count,
-            journal_len
-        );
-
-        let mut journal = self.journal.write().expect(
-            "Cannot acquire write lock on journal",
-        );
-
-        for _ in 0..count {
-            if let Some(patch) = journal.pop() {
-                self.inner.merge(patch).expect(
-                    "Cannot merge roll-back patch",
-                );
-            } else {
-                panic!("Cannot rollback the database, the journal is empty");
-            }
-        }
-
-        journal_len < count
+    pub fn rollback(&self, count: usize) {
+        self.inner
+            .write()
+            .expect("Cannot lock CheckpointDb for rollback")
+            .rollback(count);
     }
+}
 
-    /// Returns a handler to the database. The handler could be used to roll the database back
-    /// without having the ownership to it.
-    pub fn handler(&self) -> CheckpointDbHandler<T> {
-        CheckpointDbHandler(Clone::clone(self))
-    }
+#[derive(Debug)]
+struct CheckpointDbInner<T> {
+    db: T,
+    journal: Vec<Patch>,
 }
 
-impl<T: Database + Clone> Database for CheckpointDb<T> {
-    fn clone(&self) -> Box<Database> {
-        Box::new(Clone::clone(self))
+impl<T: Database> CheckpointDbInner<T> {
+    fn new(db: T) -> Self {
+        CheckpointDbInner {
+            db,
+            journal: Vec::new(),
+        }
     }
 
     fn snapshot(&self) -> Box<Snapshot> {
-        self.inner.snapshot()
+        self.db.snapshot()
     }
 
     fn merge(&mut self, patch: Patch) -> StorageResult<()> {
-        let snapshot = self.inner.snapshot();
-        self.inner.merge(patch.clone())?;
-        let mut rev_fork = self.fork();
+        // XXX: make sure that **both** the db and the journal
+        // are updated atomically.
```

Comment:
I'm not sure I understand this comment. Correct me if I'm wrong, but `CheckpointDb::merge` guarantees the required atomicity with the help of a `RwLock`. Do you mean that `CheckpointDbInner` must guarantee atomicity on its own? I'm not sure it's necessary, especially given that it's not a public type.

Reply:
The problem is visible in the old version of `merge`: https://github.com/exonum/exonum-testkit/blob/73cb36995f0dc2117291dd5e441d0688383daedc/src/checkpoint_db.rs#L109

If two patches, `A` and `B`, are merge concurrently, the database may apply them in `A, B` order, but the log might get `B, A` order .

Reply:
To expand on this a bit:

I don't think this is a super important bug to fix, because you really have to try hard to hit this race condition. 

>Do you mean that CheckpointDbInner must guarantee atomicity on its own?

Like, `CheckpointDb` as a whole must guarantee merge atomicity, and in this implementation atomicity is guaranteed by the `&mut` on merge of `Inner`. 

However, it is possible to refactor code in such a way that log and the database are protected by different locks (like was implemented originally), and this refactoring may accidentally break atomicity, which is not naturally enforced by types. So, the comment here is to remind of this implicit invariant. Does it make sense? :)